### PR TITLE
Add tensor.to(device) method.

### PR DESCRIPTION
**Pull request by gchanan**
These also take an optional dtype and behave like type conversion, i.e. if the device and dtype match, it is a no-op, otherwise there is a copy.

### CODE REVIEW

**Comment by gchanan** [comment]
I'm also open to better names; `on` seems pretty short and convenient, although the standard pattern for `on_as` is awkward, so I went with `on_device_as`.

**Comment by gchanan** [comment]
I like `device` but `device` is already an attribute and you can't have both an attribute and a function with the same name.

**Comment by soumith** [comment]
can we change the attribute to something else, like `_device`?
I am thinking about which will be used more, and I think the device-mover will be used more rather than attribute.

I think the common pattern that'd emerge is:

```
device = "cpu"
if torch.cuda.is_available():
    device = "gpu:0"

my_tensor, my_labels = DataIterator(...)

my_tensor.device(device), my_labels.device(device)
```

i.e. the device movement call will be quite prevalent.
So reducing it's verbosity is important

**Comment by gchanan** [comment]
Well, an underscore prefix has meant "not part of public API" (and tensorflow uses `device` for roughly what we are using it for now).

I'm not really a fan of using `of` instead of `as`, because it doesn't match the rest of our API, i.e. the function name is `type_as` not `type_of`.

**Comment by gchanan** [comment]
Also it's kind of nice for the property to match the parameter name, i.e.
`torch.randn(shape, device=other.device)`

**Comment by gchanan** [comment]
We get around this on `type` by making everything a function, i.e. `type()` returns the type, `type(other_type)` converts the type, `type_as(tensor)` converts the type based on tensor.  But that's kind of ugly too.

File: test/test_torch.py [hunk 1]
```diff
@@ -1510,6 +1510,36 @@ def assertEqual(device_str, fn):
...
                 assertEqual('cuda:1', lambda: torch.tensor(5, dtype=torch.int64, device='cuda:1'))
 
+    def test_on(self):
+        a = torch.tensor(5)
+        self.assertEqual(a.device, a.on('cpu').device)
+        self.assertEqual(a.device, a.on('cpu', dtype=torch.float32).device)
+        self.assertIs(torch.float32, a.on('cpu', dtype=torch.float32).dtype)
+
+        if torch.cuda.is_available():
+            for cuda in ['cuda', 'cuda:0' if torch.cuda.device_count() == 0 else 'cuda:1']:
```
Comment (SsnL) [inline test/test_torch.py]: I think this should be 
```python
'cuda:0' if torch.cuda.device_count() == 1 else 'cuda:1'
```

**Comment by apaszke** [comment]
How about we just do `to_device` and `to_device_as`? I'd say `on` is a bit ambiguous.

**Comment by gchanan** [comment]
to_device works for me.

**Comment by fmassa** [comment]
What about `astype`, like in numpy? Maybe the meaning of device gets a bit hidden, but might sound natural to users coming from numpy

**Comment by gchanan** [comment]
I think that's the wrong direction, we split dtype and device to separate those concerns.

**Comment by soumith** [comment]
After a brief offline discussion, this seems like a good idea:

```
x.to(device)
x.to(device, dtype)
x.to(dtype)
x.to(tensor) # moves it to tensor's device and changes it to tensor's dtype
x.to(tensor.device) # moves it to tensor's device
```

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Fortunately `to` is not a C++ reserved keyword :D

**Pull request merged**